src/server.cpp

```diff
@@ -1533,21 +1533,27 @@ void Server::SendNodeDef(session_t peer_id,
 	Non-static send methods
 */
 
-void Server::SendInventory(PlayerSAO* playerSAO)
+void Server::SendInventory(PlayerSAO *sao, bool incremental)
 {
-	UpdateCrafting(playerSAO->getPlayer());
+	RemotePlayer *player = sao->getPlayer();
+
+	// Do not send new format to old clients
+	incremental &= player->protocol_version >= 38;
+
+	UpdateCrafting(player);
 
 	/*
 		Serialize it
 	*/
 
-	NetworkPacket pkt(TOCLIENT_INVENTORY, 0, playerSAO->getPeerID());
+	NetworkPacket pkt(TOCLIENT_INVENTORY, 0, sao->getPeerID());
 
-	std::ostringstream os;
-	playerSAO->getInventory()->serialize(os);
-
-	std::string s = os.str();
+	std::ostringstream os(std::ios::binary);
+	sao->getInventory()->serialize(os, incremental);
+	sao->getInventory()->setModified(false);
+	player->setModified(true);
```

Comment:
why is the player marked as modified here?

Reply:
I was wondering the same thing.

Reply:
Inventories can be per-node or per-player. In this case we need to set the player dirty to save the inventory.

Reply:
Oh ok, thanks for clarifying

Reply:
But shouldn't the player be marked as modified when the inventory changes, not when it's sent?
Or does that happen at the same time anyway?

Reply:
`Server::handleCommand_InventoryAction` calls `Server::setInventoryModified`, which calls `Server::SendInventory`, which in turn calls `Player::setModified`. They do seem to happen at the same time.

https://github.com/minetest/minetest/blob/a0527851c95505c003c99a35fcc57c61c1d374d4/src/network/serverpackethandler.cpp#L558-L736

https://github.com/minetest/minetest/blob/a0527851c95505c003c99a35fcc57c61c1d374d4/src/server.cpp#L1243-L1263

Reply:
Added in both places to ensure the player is marked as modified.